cpp/plugins/cucim.kit.cuslide/src/cuslide/cuslide.cpp

```diff
@@ -82,15 +82,20 @@ static bool CUCIM_ABI checker_is_valid(const char* file_name, const char* buf, s
     return false;
 }
 
-static CuCIMFileHandle CUCIM_ABI parser_open(const char* file_path)
+static CuCIMFileHandle_share CUCIM_ABI parser_open(const char* file_path)
 {
     auto tif = new cuslide::tiff::TIFF(file_path, O_RDONLY);
     tif->construct_ifds();
-    return tif->file_handle();
+    // Move the ownership of the file handle object to the caller (CuImage).
+    auto handle_t = tif->file_handle();
+    tif->file_handle() = nullptr;
```

Comment:
The method (`tif->file_handle()`) was used for holding shared_ptr object temporary.

Plugins(libcuslide.so) are using C-API and they open a file descriptor.
CuImage objects in main library (libcucim.so) may share the file descriptor so needed to handle by using shared_ptr.

CuImage class wants to use the shared file descriptor (std::shared_ptr<CuCIMFileHandle>) created from the plugin code so that CuImage object can close the file descriptor when no objects are using the file descriptor.

```
    // Move the ownership of the file handle object to the caller (CuImage).
    auto handle_t = tif->file_handle();    // (1)
    tif->file_handle() = nullptr;               // (2)
    CuCIMFileHandle_share handle = new std::shared_ptr<CuCIMFileHandle>(handle_t);   // (3)
    return handle;                                                                                                                  // (4)
```
(1) Assigning to handle_t would increase the reference count of the shared object (`CuCIMFileHandle` object). Now `CuCIMFileHandle` object has a reference count of 2.
(2) By assigning `nullptr` to tif->file_handle(), plugin code lose the access to the shared pointer so the reference count is 1 (owned by only `handle_t` local variable).
(3) Create a `pointer` whose type is `std::shared_ptr<CuCIMFileHandle>*` (dynamic allocation) and 
(4) return the handle

The returned pointer is restored to `std::shared_ptr<CuCIMFileHandle>` from CuImage class below:

https://github.com/gigony/cucim/blob/622f5bb579d50db41973bad84c435d2968364ca4/cpp/src/cuimage.cpp#L154
```
        std::shared_ptr<CuCIMFileHandle>* file_handle_shared =
            reinterpret_cast<std::shared_ptr<CuCIMFileHandle>*>(image_format_->image_parser.open(path.c_str()));
        file_handle_ = *file_handle_shared;
        delete file_handle_shared;
```

This answers a question here: https://github.com/rapidsai/cucim/pull/191#discussion_r792436016

Moving shared_ptr between C API and C++ API has a kind of boilerplate code and it can be simplified by using a template though. (Please teach me if you know a better/fancy way to solve the issue).

The below link is also talking about the same issue:
https://www.reddit.com/r/cpp_questions/comments/d2owlo/how_can_i_pass_shared_ownership_through_a_void/


https://github.com/gigony/cucim/blob/622f5bb579d50db41973bad84c435d2968364ca4/cpp/plugins/cucim.kit.cuslide/src/cuslide/tiff/tiff.h#L71
```
    std::shared_ptr<CuCIMFileHandle>& file_handle(); /// used for moving the ownership of the file handle to the caller.
                                                     /// Do not use for the application -- it will return nullptr.
```